tower-util/src/combinators/with.rs

```diff
@@ -0,0 +1,70 @@
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pub struct With<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> With<S, F> {
+    pub fn new(inner: S, f: F) -> Self {
+        With { inner, f }
+    }
+}
+
+impl<S, F, NewRequest, OldRequest> Service<NewRequest> for With<S, F>
+where
+    S: Service<OldRequest>,
+    F: FnOnce(NewRequest) -> OldRequest,
```

Comment:
Nit: This naming feels a little backwards to me. Since this service is called _before_ the inner service, it seems like this service's `Request` type is the "old" request, and the returned request is the "new" request. I could be convinced otherwise.

Reply:
I see, my rationale here was that `S` is your "old" service and `With<S, F>` is your "new" service and hence `NewRequest` is the request for the "new" service.